src/t_zset.c

```diff
@@ -328,54 +329,64 @@ int zslIsInRange(zskiplist *zsl, zrangespec *range) {
 
     /* If everything is out of range, return early. */
     if (!zslIsInRange(zsl,range)) return NULL;
 
     x = zsl->header;
-    for (i = zsl->level-1; i >= 0; i--) {
-        /* Go forward while *OUT* of range. */
-        while (x->level[i].forward &&
-            !zslValueGteMin(x->level[i].forward->score,range))
+    if (n >= 0) {
+        for (i = zsl->level - 1; i >= 0; i--) {
+            /* Go forward while *OUT* of range. */
+            while (x->level[i].forward && !zslValueGteMin(x->level[i].forward->score, range)) {
+                /* Count the rank of the first element in range. */
+                rank += x->level[i].span;
                 x = x->level[i].forward;
-    }
-
-    /* This is an inner range, so the next node cannot be NULL. */
-    x = x->level[0].forward;
-    serverAssert(x != NULL);
-
-    /* Check if score <= max. */
-    if (!zslValueLteMax(x->score,range)) return NULL;
-    return x;
-}
-
-/* Find the last node that is contained in the specified range.
- * Returns NULL when no element is contained in the range. */
-zskiplistNode *zslLastInRange(zskiplist *zsl, zrangespec *range) {
-    zskiplistNode *x;
-    int i;
-
-    /* If everything is out of range, return early. */
-    if (!zslIsInRange(zsl,range)) return NULL;
-
-    x = zsl->header;
-    for (i = zsl->level-1; i >= 0; i--) {
-        /* Go forward while *IN* range. */
-        while (x->level[i].forward &&
-            zslValueLteMax(x->level[i].forward->score,range))
+            }
+        }
+        /* Check if zsl is long enough. */
+        if ((unsigned long)(rank + n) >= zsl->length) return NULL; 
+        if (n < ZSKIPLIST_MAX_SEARCH) {
+            /* If offset is small, we can just jump node by node */
+            for (i = 0; i < n + 1; i++) { 
+                x = x->level[0].forward; 
+            }
+        } else {
+            /* If offset is big, we jump to the nth node by its rank from the start of skiplist. */
+            x = zslGetElementByRank(zsl, rank + n + 1);
+        }
+        /* Check if score <= max. */
+        if (x && !zslValueLteMax(x->score,range)) return NULL;
+    } else  {
+        for (i = zsl->level - 1; i >= 0; i--) {
+            /* Go forward while *IN* range. */
+            while (x->level[i].forward && zslValueLteMax(x->level[i].forward->score, range)) {
+                /* Count the rank of the last element in range. */
+                rank += x->level[i].span;
                 x = x->level[i].forward;
+            }
+        }
+        /* Check if the range is big enough. */
+        if (rank < -n) return NULL;
+        if (n + 1 > -ZSKIPLIST_MAX_SEARCH) {
+            /* If offset is small, we can just jump node by node */
+            for (i = 0; i < -n - 1; i++) {
+                x = x->backward;
+            }
+        } else {
+            /* If offset is big, we jump to the nth node by its rank from the start of skiplist. */
+            x = zslGetElementByRank(zsl, rank + n + 1);
```

Comment:
i noticed that in the old code (`zslLastInRange`), we didn't used to go one extra step (unlike what we did in zslFirstInRange).
i.e. 
```
    /* This is an inner range, so the next node cannot be NULL. */
    x = x->level[0].forward;
```
vs
```
    /* This is an inner range, so this node cannot be NULL. */
```
is this solved by `n` being `-1` in this version?
(sorry for my lack of focus)

Reply:
it's better to add more comments... since the different bases: rank(1-based), zslGetElementByRank(1-based), n(0-based when forward), n(-1-based when backward).

Reply:
> is this solved by `n` being `-1` in this version?

Yes. And I just added some comments to illustrate `rank` and `n`.